version: '3.7'

services:
  auth:
    image: dnikolaidis/disastergram-auth:nginx
    build:
      context: .
      dockerfile: Dockerfile-nginx
    deploy:
      replicas: 2
      restart_policy:
        condition: on-failure
    environment:
      POSTGRES_HOST: auth-db
      POSTGRES_USER: postgres
      POSTGRES_PASSWORD: disastergram
      BASEURL: http://localhost:5000/
      DOCKER_HOST: auth:5000
      NGINX_WORKER_PROCESSES: 4
      FLASK_APP_INSTANCE: /app/instance
      FLASK_APP: /app/auth
    ports:
    - "5000:80"
    depends_on:
      - auth-db
      - zoo1
      - zoo2
      - zoo3

  # https://hub.docker.com/_/postgres
  auth-db:
    image: postgres:11
    restart: always
    environment:
      POSTGRES_PASSWORD: disastergram
    volumes:
      - auth-data:/var/lib/postgresql/
      
  zoo1:
    image: zookeeper
    restart: always
    hostname: zoo1
    ports:
      - 2181:2181
    environment:
      ZOO_MY_ID: 1
      ZOO_SERVERS: server.1=0.0.0.0:2888:3888 server.2=zoo2:2888:3888 server.3=zoo3:2888:3888
    
  zoo2:
    image: zookeeper
    restart: always
    hostname: zoo2
    ports:
      - 2182:2181
    environment:
      ZOO_MY_ID: 2
      ZOO_SERVERS: server.1=zoo1:2888:3888 server.2=0.0.0.0:2888:3888 server.3=zoo3:2888:3888
    
  zoo3:
    image: zookeeper
    restart: always
    hostname: zoo3
    ports:
      - 2183:2181
    environment:
      ZOO_MY_ID: 3
      ZOO_SERVERS: server.1=zoo1:2888:3888 server.2=zoo2:2888:3888 server.3=0.0.0.0:2888:3888

volumes:
  auth-data: